docs: Polish the langref concepts (#2241)
slint-ui/slint: docs/langref/src/concepts/layouting.md

# Positioning and Layout of Elements

All visual elements are shown in a window. The `x` and `y` properties store
the elements coordinates relative to their parent element. You calculate the
absolute position of an element in a window by adding the parent's position to
the element's position. If the parent has a grandparent element, then that one
is added as well. This calculation continues until the top-level element

> is reached.

Reviewer: Why did you rephrase it as "You calculate"? This implies that the developer has some kind of obligation. I prefer the passive form.

Another way would be to write "Slint determines the absolute position of an element by ..."

The `width` and `height` properties store the size of visual elements.

You can create an entire graphical user interface by placing the elements in
two ways:

* Explicitly - by setting the `x`, `y`, `width`, and `height` properties.
* Automatically - by using layout elements.

> Explicit placement is great for static scenes with few elements. Layouts are
> suitable for complex user interfaces as they handle many elements at one in
> an automated way. Layouts express geometric relationship between elements in
> dedicated layout elements.

Reviewer: I think "scalability" is a better argument for layouts than "handle many elements at one in an automated way".

```suggestion
Explicit placement is great for static scenes with few elements. Layouts are
suitable for complex user interfaces and help create scalable user interfaces. Layout elements express geometric relationships between elements.
```

}
```

> The positions of both rectangles and the size of the inner green one are hard-coded.

Reviewer: I think "fixed" is a much better counter part to variable than "hard-coded", a term that I think is primarily used for code, while this here is about layouts.

```suggestion
The positions of both rectangles and the size of the inner green one are fixed.
```

-   `Rectangle`
-   `TouchArea`
-   `FocusScope`
-   `Flickable`

## Automatic Placement Using Layouts

Slint comes with different layout elements that automatically calculate the position and size of their children:

-   `VerticalLayout` / `HorizontalLayout`: The children are placed along the vertical or horizontal axis.
-   `GridLayout`: The children are placed in a grid of columns and rows.

You can also nest layouts, making it possible to create complex user interfaces.

You can tune the automatic placement using different constraints, to accommodate the design of your user
interface. For example each element has a minimum and a maximum size. Set these explicitly using the
following properties:

-   `min-width`
-   `min-height`
-   `max-width`
-   `max-height`

A layout element also affects the minimum and maximum size of its parent.

Any element with a specified `width` and `height` has a fixed size in a layout.

When there is extra space in a layout, elements can stretch along the layout axis. You can control this stretch
factor between the element and its siblings with these properties:

-   `horizontal-stretch`
-   `vertical-stretch`

> A value of `0` means that the element won't stretch at all, unless all siblings
> also have a stretch factor of `0`. In this case all elements stretch equally.

Reviewer: It's still a little unclear if "In this case" refers to "A value of `0`" or "all siblings have a stretch factor of `0`". How about this?

```suggestion
A value of `0` means that the element won't stretch at all. All elements stretch equally if they all have a stretch factor of `0`.
```

A layout element also affects the minimum and maximum size of its parent.

> Any element with a specified `width` and `height` has a fixed size in a layout.

Reviewer: Nice simplification, and also a consistent use of "fixed" instead of say "has a hard-coded size in a layout" :-)

When there is extra space in a layout, elements can stretch along the layout axis. You can control this stretch
factor between the element and its siblings with these properties:

* `horizontal-stretch`
* `vertical-stretch`

A value of `0` means that the element won't stretch at all, unless all siblings
also have a stretch factor of `0`. In this case all elements stretch equally.

The default value of these constraint properties may depends on the content of the element.

> If the element doesn#t set a `x` or a `y` property, these constraints are also automatically applied to the parent element.

Reviewer: The element doesn't set properties, it's the user that sets them :-)

```suggestion
If the element's `x` or `y` is not set, these constraints are also automatically applied to the parent element.
```

## `VerticalLayout` and `HorizontalLayout`

> The `VerticalLayout` and `HorizontalLayout` elements place elements in a column or a row.

Reviewer: ```suggestion
The `VerticalLayout` and `HorizontalLayout` elements place children in a column or a row.
```

> By default, they will be stretch or shrink so that they take the whole space. You
> can adjust the element's alignmnet as needed.

Reviewer: ```suggestion
By default, they will be stretched or shrunk so that they take the whole space. You
can adjust the element's alignment as needed.
```

or

```suggestion
By default, they stretch or shrink so that they take the whole space. You
can adjust the element's alignment as needed.
```

}
```

> The example below, on the other hand, specifies that the rectangles shall be aligned

Reviewer: Maybe even simpler :-)

```suggestion
The example below, on the other hand, specifies that the rectangles shall align
```

### Alignment

> Each elements is sized according to their `width` or `height` if specified, otherwise it is

Reviewer: ```suggestion
Each element is sized according to their `width` or `height` if specified, otherwise it is
```

set to the minimum size which is set with the min-width or min-height property, or
the minimum size of an inner layout, whatever is bigger.

Then, the elements are placed according to the alignment. The size of elements

> is bigger than the minimum size only if the alignment is stretch

Reviewer: "alignment is stretch" doesn't make sense to me (but was there before). @ogoffart Do you remember what is meant here? Should it perhaps be "only if the layout stretches it"?

Reviewer 2: ```suggestion
The elements are placed according to the alignment. The size of elements
is bigger than the minimum size only if the `alignment` property of the layout is `LayoutAlignment.stretch` (the default)
```
